### PR TITLE
Verbose client

### PR DESCRIPTION
**Pull request by granders**
Adds "verbose producer" client to help with system testing.

I didn't add a corresponding consumer since console consumer was good enough for my purposes.

### CODE REVIEW

**Comment by ghost** [comment]
Hey @granders,
thank you for your Pull Request.

It looks like you haven't signed our **C**ontributor **L**icense **A**greement, yet.

> The purpose of a CLA is to ensure that the guardian of a project's outputs has the necessary ownership or grants of rights over all contributions to allow them to distribute under the chosen licence.
> [Wikipedia](http://en.wikipedia.org/wiki/Contributor_License_Agreement)

You can read and sign our full Contributor License Agreement [here](http://clabot.confluent.io/cla).

Once you've signed reply with `[clabot:check]` to prove it.

Appreciation of efforts,

clabot

File: core/src/main/scala/kafka/tools/ProducerPerformance.scala [hunk 1]
```diff
@@ -115,9 +115,9 @@ object ProducerPerformance extends Logging {
       .defaultsTo(0)
     val csvMetricsReporterEnabledOpt = parser.accepts("csv-reporter-enabled", "If set, the CSV metrics reporter will be enabled")
     val metricsDirectoryOpt = parser.accepts("metrics-dir", "If csv-reporter-enable is set, and this parameter is" +
-      "set, the csv metrics will be outputed here")
+      "set, the csv metrics will be output here")
```
Comment (ewencp) [inline core/src/main/scala/kafka/tools/ProducerPerformance.scala:118]: These look unrelated? Probably not a big deal since this patch is going to Kafka and this is just some cleanup, but wanted to make sure this was intentional.

File: core/src/main/scala/kafka/tools/VerboseProducer.java [hunk 2]
```diff
@@ -0,0 +1,185 @@
+package kafka.tools;
```
Comment (ewencp) [inline core/src/main/scala/kafka/tools/VerboseProducer.java]: Seems like this is in a weird location? It's a java file under a scala/ directory for one. Also, any reason this is under core/ given that it uses the new producer? Seems like it could be placed under clients/src/main/java/org/apache/kafka/clients/tools/ along with the new ProducerPerformance tool.

File: core/src/main/scala/kafka/tools/VerboseProducer.java [hunk 3]
```diff
@@ -0,0 +1,185 @@
...
+
+import joptsimple.ArgumentAcceptingOptionSpec;
+import joptsimple.OptionParser;
+import joptsimple.OptionSet;
+import joptsimple.OptionSpec;
+import joptsimple.OptionSpecBuilder;
+import kafka.utils.CommandLineUtils;
+
+/**
+ * Primarily intended for use with system testing, this producer produces a fixed number of
```
Comment (ewencp) [inline core/src/main/scala/kafka/tools/VerboseProducer.java]: This comment is a bit misleading since the "fixed number of increasing integers" is in the static main method, not the class itself. Is the key property that it logs metadata in JSON to stdout, allowing for validation? The fixed # of increasing IDs may be a specific use case.

Figuring out what that key property is might help with naming the class.

File: core/src/main/scala/kafka/tools/VerboseProducer.java [hunk 4]
```diff
@@ -0,0 +1,185 @@
...
+import joptsimple.OptionSpec;
+import joptsimple.OptionSpecBuilder;
+import kafka.utils.CommandLineUtils;
+
+/**
+ * Primarily intended for use with system testing, this producer produces a fixed number of 
+ * increasing integers and prints metadata in the form of JSON to stdout for each failed or 
+ * successful produce attempt.
+ */
+public class VerboseProducer {
```
Comment (ewencp) [inline core/src/main/scala/kafka/tools/VerboseProducer.java]: Any ideas for clearer names for the class? VerboseProducer makes it sound like this could be replaced by just adding a --verbose flag to something like producer performance.

File: core/src/main/scala/kafka/tools/VerboseProducer.java [hunk 5]
```diff
@@ -0,0 +1,185 @@
...
+  OptionParser commandLineParser;
+  Map<String, OptionSpec<?>> commandLineOptions = new HashMap<String, OptionSpec<?>>();
+  
+  String topic;
+  String sync;
```
Comment (ewencp) [inline core/src/main/scala/kafka/tools/VerboseProducer.java]: Looks like this isn't used. Leftover from previous version of the code?

File: core/src/main/scala/kafka/tools/VerboseProducer.java [hunk 6]
```diff
@@ -0,0 +1,185 @@
...
+    commandLineOptions.put("help", helpOpt);
+  }
+  
+  /** Validate command-line arguments and parse into properties object. */
+  public void parseCommandLineArgs(String[] args) throws IOException {
+    
+    OptionSpec[] requiredArgs = new OptionSpec[]{commandLineOptions.get("topic"),
+                                                 commandLineOptions.get("broker-list"),
+                                                 commandLineOptions.get("num-messages")};
+    if(args.length == 0) {
```
Comment (ewencp) [inline core/src/main/scala/kafka/tools/VerboseProducer.java]: This test is technically valid now, but given the check for required args later, you can just omit this. There's an outstanding patch to fix this in a lot of other command line tools in Kafka.

File: core/src/main/scala/kafka/tools/VerboseProducer.java [hunk 7]
```diff
@@ -0,0 +1,185 @@
...
+  }
+
+  /** 
+   * Return JSON string encapsulating basic information about the exception, as well
+   * as the value which triggered the exception.
+   */
+  String errorString(Exception e, String value) {
+    return "{\"producer\": \"VerboseProducer\", " 
+           + "\"exception\": \"" + e.getClass() + "\"," 
+           + "\"message\": \"" + e.getMessage() + "\","
```
Comment (ewencp) [inline core/src/main/scala/kafka/tools/VerboseProducer.java]: I think everything else is newline-safe, but messages may contain newlines. You might need to do something to make sure they are escaped and/or removed to make this robust.

File: core/src/main/scala/kafka/tools/VerboseProducer.java [hunk 8]
```diff
@@ -0,0 +1,185 @@
...
+  private class PrintInfoCallback implements Callback {
+    private String value;
+    
+    PrintInfoCallback(String value) {
+      this.value = value;
+    }
+    
+    public void onCompletion(RecordMetadata recordMetadata, Exception e) {
+      if (e == null) {
+        System.out.println(successString(this.value, recordMetadata));
```
Comment (ewencp) [inline core/src/main/scala/kafka/tools/VerboseProducer.java]: If this is logging to stdout, is the expectation that this will be run with all other logs turned off? Or just that lines that don't parse as JSON will be ignored?

File: core/src/main/scala/kafka/tools/ProducerPerformance.scala (hunk 1 above)
Comment (granders) [inline core/src/main/scala/kafka/tools/ProducerPerformance.scala:118]: Just minor cleanup - the typos were bugging me :)

File: core/src/main/scala/kafka/tools/VerboseProducer.java (hunk 3 above)
Comment (granders) [inline core/src/main/scala/kafka/tools/VerboseProducer.java]: Yes, the key property is that it logs metadata in JSON to stdout

File: core/src/main/scala/kafka/tools/VerboseProducer.java (hunk 8 above)
Comment (granders) [inline core/src/main/scala/kafka/tools/VerboseProducer.java]: I'm expecting non-JSON to be ignored. Added comment to reflect this.

**Comment by granders** [comment]
@ewencp Ok I hit your comments. I added a few dependencies to 'clients' for parsing command line options and encoding json, but I'm not sure whether it's kosher to do so.

File: build.gradle [hunk 9]
```diff
@@ -352,6 +352,8 @@ project(':clients') {
     compile "org.slf4j:slf4j-api:1.7.6"
     compile 'org.xerial.snappy:snappy-java:1.1.1.6'
     compile 'net.jpountz.lz4:lz4:1.2.0'
+    compile 'net.sf.jopt-simple:jopt-simple:4.8'
```
Comment (ewencp) [inline build.gradle]: I think we'll just need feedback from people eventually on how much they care about the additional dependencies. This might be a good reason for splitting up the tools into a separate artifact -- leave kafka-clients for just the code you'd link to when using it in your own code and then bundle all the tools into a separate kafka-client-tools jar, which are generally going to be used by people operating (or testing) Kafka.

**Comment by ewencp** [comment]
@granders Looks good so far. Not crazy about the new name either, but it's at least clearer than the previous one :) Two more things:
1. Wrapper script? I tested by running with kafka-run-class.sh, which works fine but requires knowing the full class name. Is it worth having the trivial wrapper script?
2. It's likely we'll need some sort of rate limiting. The types of tests this is useful with require running for awhile, e.g. to give enough time to bounce the servers. It may not make sense to generate messages full bore.

**Comment by ewencp** [comment]
Oops, also another followup: doesn't pass checkstyle, which makes it impossible to run tests. Indentation in Kafka doesn't use the same style as Confluent's.

Needs:
1. Four-space indentation (use IntelliJ's default formatting)
2. License header
3. Fix \* imports to only import classes that are actually used.
4. Update import-control.xml to work with new libraries.

File: clients/src/main/java/org/apache/kafka/clients/tools/MetadataToStdoutProducer.java [hunk 10]
```diff
@@ -0,0 +1,195 @@
...
+  /**
+   * Produce a message with given value and no key.
+   */
+  public void send(String key, String value) {
+    ProducerRecord<String, String> record = new ProducerRecord<String, String>(topic, key, value);
+    try {
+      producer.send(record, new PrintInfoCallback(key, value));
+    }
+    catch (Exception e) {
+      System.out.println(errorString(e, key, value));
```
Comment (ewencp) [inline clients/src/main/java/org/apache/kafka/clients/tools/MetadataToStdoutProducer.java]: This println and the one in the callback are happening on different threads. Any idea whether these calls are synchronized on different JDK implementations since there seems to be no guarantee by the PrintStream class? My guess would be that it is on most implementations, but we still might need to synchronize these statements to be sure. Otherwise garbled output could make it impossible to parse the output correctly and lead to invalid system test results.

**Comment by granders** [comment]
- Added synchronization on println (protect against interleaved prints)
- Fixed checkstyle
- Added rate limiting
- Changed name to VerifiableProducer

File: build.gradle [hunk 11]
```diff
@@ -352,6 +352,8 @@ project(':clients') {
...
+    compile 'com.googlecode.json-simple:json-simple:1.1.1'
```
Comment (ewencp) [inline build.gradle:356]: We probably don't need to change anything now since this is something other people might care about more than we do, but people may prefer jackson instead since it's used elsewhere in the code already (at least in core, maybe elsewhere).

File: clients/src/main/java/org/apache/kafka/clients/tools/VerifiableProducer.java [hunk 12]
```diff
@@ -0,0 +1,279 @@
...
+    public VerifiableProducer(String[] args) throws IOException {
+        this.configureParser();
+        this.parseCommandLineArgs(args);
+        this.producer = new KafkaProducer<String, String>(producerProps);
+    }
+  
+    /** Set up the command-line options. */
+    private void configureParser() {
+        this.commandLineParser = new OptionParser();
+        ArgumentAcceptingOptionSpec<String> topicOpt = commandLineParser.accepts("topic", "REQUIRED: The topic id to produce messages to.")
```
Comment (ewencp) [inline clients/src/main/java/org/apache/kafka/clients/tools/VerifiableProducer.java]: Have you kept track of KIP-4? I think there was some discussion of doing a better job of standardizing command line arguments in the tools. Might be worth a) double checking it to make sure nothing here would be inconsistent (unlikely since it's about admin commands) and b) check the mailing list to see if there was any discussion of argument parsing which might end up affecting this code, e.g. if they had discussed/decided on which library to use.

File: clients/src/main/java/org/apache/kafka/clients/tools/VerifiableProducer.java [hunk 13]
```diff
@@ -0,0 +1,279 @@
...
+        this.numMessages = (Integer) options.valueOf("num-messages");
+        this.topic = (String) options.valueOf("topic");
+        this.throughput = (Long) options.valueOf("throughput");
+    
+        producerProps.put(ProducerConfig.BOOTSTRAP_SERVERS_CONFIG, options.valueOf("broker-list"));
+        producerProps.put(ProducerConfig.KEY_SERIALIZER_CLASS_CONFIG,
+                          "org.apache.kafka.common.serialization.StringSerializer");
+        producerProps.put(ProducerConfig.VALUE_SERIALIZER_CLASS_CONFIG,
+                          "org.apache.kafka.common.serialization.StringSerializer");
+        producerProps.put(ProducerConfig.ACKS_CONFIG, "-1");
```
Comment (ewencp) [inline clients/src/main/java/org/apache/kafka/clients/tools/VerifiableProducer.java]: This seems like one of the key properties you might want to be able to configure. Alternatively, should we be able to override any producer property via the command line, except for maybe things like `retries`?

File: clients/src/main/java/org/apache/kafka/clients/tools/VerifiableProducer.java [hunk 14]
```diff
@@ -0,0 +1,279 @@
...
+                }
+            }
+        }
+    }
+  
+    public static void main(String[] args) throws IOException {
+    
+        VerifiableProducer producer = new VerifiableProducer(args);
+
+        long sleepTimeNs = NS_PER_SEC / producer.throughput;
```
Comment (ewencp) [inline clients/src/main/java/org/apache/kafka/clients/tools/VerifiableProducer.java]: This duplicates some other code pretty closely, and in Confluent's code I tried to make this reusable. Does that make sense here? Minor duplication is fine, but this seems like a pretty sizable chunk that's duplicated. Then again, refactoring the code in a way to make it reusable eventually got kind of messy in Confluent's code...

**Comment by ewencp** [comment]
@granders Looks good, and simple test worked perfectly. Remaining comments aren't major, but some might need to be addressed in a subsequent Kafka review anyway. I'll leave it up to you whether you want to defer those (especially the speculative ones) or address them now.

File: clients/src/main/java/org/apache/kafka/clients/tools/VerifiableProducer.java (hunk 12 above)
Comment (granders) [inline clients/src/main/java/org/apache/kafka/clients/tools/VerifiableProducer.java]: Good point, I double-checked the KIP and
a) From my vantage point, the interface is consistent/similar in feel to kafka.sh
b) I did some superficial searching of the dev list but didn't find argument parsing discussions

**Comment by granders** [comment]
I shared throttling logic by adding MessageThroughputThrottler class to the clients.tools directory, let me know what you think of this approach.

**Comment by ewencp** [comment]
@granders Looks like a good refactoring, might just reduce the name -- it's not really MessageSpecific. Just "Throttler" might be descriptive enough.

**Comment by granders** [comment]
@ewencp Oh, I chose the name since it regulates the average rate at which producer.send gets called, and I was imagining a non-existent companion DataThroughputThrottler which would take message sizes into account.

**Comment by ewencp** [comment]
@granders My point is that nothing in the class itself is message-specific (except maybe the argument name to shouldThrottle). ThroughputThrottler would be generally accurate. For example, if you wanted to throttle in bytes/s rather than msg/s, you could just make the throughput passed to the constructor be in bytes/s and the "messageNum" be the aggregate number of bytes so far and it should still work.

**Comment by granders** [comment]
@ewencp Ok good point I didn't think of that.

**Comment by granders** [comment]
@ewencp 
I'll think I'll defer the other questions to patch review. Good to merge?

**Comment by granders** [comment]
I should mention that I made it possible to run the producer indefinitely.

This is more appropriate if we want the producer working the entire time while failures are being driven.
The test driver then has the option to stop the process when ready.

File: clients/src/main/java/org/apache/kafka/clients/tools/VerifiableProducer.java [hunk 15]
```diff
@@ -0,0 +1,261 @@
...
+    public static void main(String[] args) throws IOException {
+        
+        final VerifiableProducer producer = new VerifiableProducer(args);
+        final long startMs = System.currentTimeMillis();
+        boolean infinite = producer.maxMessages < 0;
+
+        Runtime.getRuntime().addShutdownHook(new Thread() {
+            @Override
+            public void run() {
+                producer.close();
```
Comment (ewencp) [inline clients/src/main/java/org/apache/kafka/clients/tools/VerifiableProducer.java:262]: Can we also set a flag to stop the main thread from trying to produce more data? At first I thought ctrl-c wasn't working because it kept logging so many messages, and then the tail of the log is a bunch of errors due to the producer being closed. It also caused the summary output to not be the final output, which is a bit weird because then the stats printed out might not be accurate.

File: clients/src/main/java/org/apache/kafka/clients/tools/VerifiableProducer.java [hunk 16]
```diff
@@ -0,0 +1,261 @@
...
+        commandLineOptions.put("max-messages", numMessagesOpt);
+
+        ArgumentAcceptingOptionSpec<String>  throughputOpt = commandLineParser.accepts("throughput", "Average message throughput, in messages/sec. Default: -1, results in no throttling.")
+                .withOptionalArg()
+                .defaultsTo("-1")
+                .describedAs("throughput")
+                .ofType(String.class);
+        commandLineOptions.put("throughput", throughputOpt);
+
+        ArgumentAcceptingOptionSpec<String>  acksOpt = commandLineParser.accepts("acks", "number of acks required. Default: -1")
```
Comment (ewencp) [inline clients/src/main/java/org/apache/kafka/clients/tools/VerifiableProducer.java]: Maybe it doesn't matter since it's the default, but I can't actually use --acks -1 because the option parser thinks the -1 part is an option.

Actually, it looks like --acks=-1 is ok, so maybe this is a quirk in option parsing we may not easily be able to fix?

**Comment by ewencp** [comment]
@granders a couple of questions, but otherwise LGTM. A --help option might also be nice (the full list of options was removed with the removal of the `args.length == 0` check).

**Comment by granders** [comment]
@ewencp Sorry to keep churning on this, but I could not accept such weird behavior wrt "-1" on the command-line. I wasn't able to figure out a way to make it work in joptsimple, so I investigated a few options and argparse4j is more robust and has cleaner api.

Added flag to stop producing in main thread when shutdown hook is triggered.
Help prints when no args are present.
Added trivial shell program.

**Comment by ewencp** [comment]
LGTM
